### PR TITLE
rbac: clean up role assignments and related AAD application when delete a service principal

### LINKED ISSUE #3043: `az ad sp create-for-rbac` fails even after deleting the created sp
### Description
It seems like `az ad sp create-for-rbac` is creating a service principal only, so after running `az ad sp delete`, deleting the sp leaves the application (which makes sense).

The problem is when deleting a sp created with create-for-rbac, then trying to run `az ad sp create-for-rbac`, the command fails unless you delete the app.

Seems this command should be `az ad app create-for-rbac` or take into account the app may still exist and simply add another sp to the existing app.

### LINKED ISSUE #3508: Update description for `role assignment list`
### Description

The description for `az role assignment list` should give details on the default values provided.  This command is confusing in use right now as you can create a role assignment against a resource group, but not see it in the assignment list (without a --all), but there are no details on what is or is not shown.  

Current:
```
az role assignment list: List role assignments.

Arguments
    --all              : Show all assignments under the current subscription.
```

I propose something like:

```
az role assignment list: List role assignments.  By default, only assignments scoped to subscription will be displayed.  To view assignments scoped by resource or group, use --all. 

Arguments
    --all              : Show all assignments under the current subscription.
```

### PR DESCRIPTION
**Pull request by yugangw-msft**
Fix #3043, Fix #3508

- [x] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 1]
```diff
@@ -479,9 +479,27 @@ def show_service_principal(client, identifier):
...
+        if result:
+            app_object_id = result[0].object_id
+
+    assignments = list_role_assignments(assignee=identifier, show_all=True)
+    if assignments:
+        logger.warning('Removing role assignments')
+        delete_role_assignments([a['id'] for a in assignments])
+
+    if app_object_id:
+        client.applications.delete(app_object_id)  # this should auto-delete service principals
```
Comment (tjprescott) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Do we test this?

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py [hunk 2]
```diff
@@ -479,9 +479,27 @@ def show_service_principal(client, identifier):
     return client.get(object_id)
 
 
-def delete_service_principal(client, identifier):
-    object_id = _resolve_service_principal(client, identifier)
-    client.delete(object_id)
+def delete_service_principal(identifier):
```
Comment (tjprescott) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Should we have a flag to retain these artifacts?  Is there any use case that would require that?

File: src/command_modules/azure-cli-role/tests/test_graph.py [hunk 3]
```diff
@@ -91,6 +91,34 @@ def body(self):
...
+        self.cmd('ad sp create-for-rbac -n ' + name)
+
+        self.cmd('ad sp list --spn ' + name)
+
+        self.cmd('ad app list --identifier-uri ' + name)
+
+        result = self.cmd('role assignment list --assignee ' + name).get_output_in_json()
+        object_id = result[0]['properties']['principalId']
+
+        self.cmd('ad sp delete --id ' + name)
```
Comment (tjprescott) [inline src/command_modules/azure-cli-role/tests/test_graph.py]: Does this scenario delete the application or the SP?  It seems like we should have a test for each (to make sure deleting the app deletes the SP)

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 1 above)
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: AAD service guarantee this behavior. Yep, I tested it as well. Let me update the code comments to reflect that

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 2 above)
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: I thought about that.  But after you delete a service principal, you will not be able to get it back with the same object id, hence the existing role assignments will become invalid which we should just clean up.

File: src/command_modules/azure-cli-role/tests/test_graph.py (hunk 3 above)
Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/tests/test_graph.py]: We do `clean delete` to remove both role assignments and the application this SP wraps. The behavior of `auto deleting application on deleting a sp` is a server behavior that we have multiple tests already leveraged, so we should be fine

File: src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py (hunk 2 above)
Comment (tjprescott) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: That makes sense for the role assignments. What about the app ID?

Comment (yugangw-msft) [inline src/command_modules/azure-cli-role/azure/cli/command_modules/role/custom.py]: Deleting a service principal and leaving the application behind in the tenant is not a scenario in rbac, not very meaningful as well in general AAD auth world. The application is useless once the SP is gone. That is why in portal, AAD UI will automatically create a SP when you author an application. 
For multi-tenants scenario(not related with rbac) that the SP was created for an external application (through admin consent), the code change here will only delete the SP. The application still stays in the original tenant.

**Pull request merged**